reps/2022-03-10-plasma_alternative.md

```diff
@@ -0,0 +1,179 @@
+print(v6d.get(v6d_object_id)) # [1,2,3,4]
+```
+
+### Lifecycle Management
+
+The goal of  lifecycle management in local plasma store is to drop some not-in-use objects (evict or spill) when we we ran out of memory. Both plasma and vineyard have thier own lifecycle management, and their behavior may be different. We think that raylet logic should not rely on the underlying lifecycle management strategies  (e.g. different eviction policy).  We propose to define a **protocol** here that both vineyard and plasma should agree. For example, the protocol will define when a object adds/remove its reference, when object can be evicted/spilled. In other word The protocol defines the impact on lifecycle management in server-side when calling the above  `ClientImplInterface` API in client-side. Since most of concerns about the integration may happens in this part, we may need further discussions about these protocols.
+
+#### Object Ref_count
+
+The Ray keeps two type of "reference count", one is `ref_count` in plasma to track the local plasma object, and the other is the `reference_count` in core_worker to track the primary copy. In this proposal, we will not change the logic of `reference_count`, and the vineyard server will have its own `ref_count`.
+
+**Protocol #1**: A object will increase its `ref_count` iff the `Get(...)` or `CreateAndSpillIfNeeded(...)`  is invoked. ([Create](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store.cc#L193), [Get](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store.cc#L106))
+
+**Protocol #2:** A object will decrease its `refer_count` iff the `release(...)` or `Disconnect(...)`is invoked. ([Release](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store.cc#L268), [DIsconnect](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store.cc#L337))
+
+#### Object eviction
+
+As mentioned above, the eviction policy may be different in different store variants. Currently, the policies of Plasma and Vineyard are both based on LRU.
+
+**Protocol #3:** Eviction only happens in an eviction request and create request. ([evict]( https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store.cc#L441), [create](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/object_lifecycle_manager.cc#L189))
+
+**Protocol #4:** Servers can only evict the objects that are not needed by clients, (a.k.a.reference count = 0). specially, pinned object should never be evcited. ([AddtoCache](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/object_lifecycle_manager.cc#L159))
+
+#### Object spilling
+
+When creating a new object with insufficient memory even with evicting possible objects, the local object manager will try to spill objects to external storage. The code path of spilling is out of Plasma (via callback) thus we can reuse it in third-party object-store.
```

Comment:
Today the object spilling works in following way:

- The underlining object store provides an [API](https://sourcegraph.com/github.com/ray-project/ray@149d06442bd0e53010ead72e4a7c0620ee1c0966/-/blob/src/ray/object_manager/object_manager.cc?L164) to query if an object is spillable.
- When we create plasma store, we need to pass a [callback](https://sourcegraph.com/github.com/ray-project/ray/-/blob/src/ray/object_manager/plasma/store_runner.h?L20) to it to allow it triggers spilling. 

We need to capture this in the storage interface, otherwise it's likely the new storage will unlikely to work.

Reply:
This API is confusing here since the invocation of this API is in the `spilling_callback`([here](https://github.com/ray-project/ray/blob/master/src/ray/raylet/local_object_manager.cc#L186)). Can we deprecate this API and make the query happen directly on the server-side (which means when store pass a list of objects to the spilling_callback, is should guarantee all the objects are spillable)?